18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,115 @@
+import React, { useCallback, useEffect, useState, useRef } from 'react';
+import io from 'socket.io-client';
+
+import Message from './components/Message';
+
+import './Chat.css';
+
+const socketUrl = 'http://localhost:3002';
+const client = io(socketUrl);
+
+function Chat() {
+  const socket = useRef(client).current;
+  const [status, setStatus] = useState('Disconnected');
+  const [inputValue, setInputValue] = useState('');
+  const [username, setUsername] = useState('anon');
+  const [messages, setMessages] = useState([]);
+  const [typingUsers, setTypingUsers] = useState(new Set());
+
+  const addMessage = useCallback((msg) => {
+    setMessages([...messages, msg]);
+  }, [messages, setMessages]);
+
+  const addTyping = useCallback((data) => {
+    setTypingUsers((typing) => new Set([...typing, data.username]));
```

Comment:
looks like, if user doesn't send a message, the typing message won't disappear.
I'd suggest adding a timeout for 2-4 secs after getting the typing event for deleting user. If you get a message about typing again you should restart timeout.

Reply:
fixed